Added Gift Card Module

mageaustralia opened the pull request:
## Description
Adds comprehensive gift card functionality to Maho ecommerce platform.

## Features
- **Gift Card Products**: Create gift card products with fixed or custom amounts
- **Admin Management**: Full admin interface for creating, editing, and managing gift cards
- **Checkout Integration**: Apply gift cards during checkout with support for partial payments
- **Balance Tracking**: Automatic balance tracking with full transaction history
- **PDF Export**: Generate printable gift cards with customizable templates
- **QR/Barcode Generation**: Local generation of QR codes and barcodes (no external API dependencies)
- **Email Notifications**: Send gift cards to recipients via email
- **Refund Support**: Handle refunds back to gift cards
- **Payment Method Filtering**: Automatically hide unnecessary payment methods when gift cards cover the full order amount

## Technical Details
- Uses BaconQrCode for QR code generation (already included in Maho)
- Uses Picqer/php-barcode-generator for barcode generation (optional dependency added)
- Fully compatible with Maho's modern JavaScript (no jQuery/Prototype dependencies)
- Includes database migrations for gift card tables
- Configurable via System > Configuration > Sales > Gift Cards

## Module Structure
- Models for gift cards, history, and scheduled emails
- Admin controllers and grids
- Frontend checkout integration
- Total collectors for quotes, orders, invoices, credit memos
- Observer events for order processing and activation
- Product type implementation for gift card products

## Gift Card PDF Customization

Gift card PDFs use the **frontend theme** of the store the gift card belongs to, following Maho's standard theme fallback system.

Default template and CSS:
```
app/design/frontend/base/default/template/maho/giftcard/pdf/
├── giftcard.phtml
└── giftcard.css
```

To customize, override in your theme:
```
app/design/frontend/[package]/[theme]/template/maho/giftcard/pdf/
├── giftcard.phtml   # for layout changes
└── giftcard.css     # for styling changes
```

Each store automatically uses its own theme, allowing different gift card designs per store in multi-store setups.

## Testing
- Tested with gift card creation, redemption, and balance tracking
- Verified PDF export with QR codes and barcodes
- Confirmed payment method filtering works correctly
- Tested refund functionality

Commit:
Add Maho Gift Card module

- Complete gift card functionality for order payments
- Support for gift card products with customizable amounts
- Admin interface for managing gift cards
- Gift card application during checkout with partial payment support
- Automatic balance tracking and history
- PDF export functionality with customizable templates
- Local QR code generation using BaconQrCode
- Local barcode generation using Picqer library
- Email notifications for gift card recipients
- Configurable settings in System > Configuration > Sales > Gift Cards
- Payment method filtering when gift cards fully cover orders
- Support for refunds back to gift cards
- Fixed minicart JavaScript error with global initialization
- Fixed Save and Continue Edit button with vanilla JavaScript

Commit:
Fix copyright notices in template files and update composer.lock

- Add proper Maho copyright headers to all gift card template files
- Update composer.lock to include picqer/php-barcode-generator dependency

Commit:
Add English translations for gift card module

- Add comprehensive translation file for all gift card module strings
- Includes translations for admin interface, frontend templates, and system messages

Commit:
Add missing translations and update PHPStorm metadata

- Add remaining missing translations for gift card module
- Regenerate PHPStorm metadata files for IDE support

Commit:
Sort gift card translation CSV alphabetically

- Sort translations alphabetically to meet CI requirements
- Fix formatting for consistency with other translation files

Commit:
Fix CI issues: code style, translations, and metadata

- Apply PHP CS Fixer formatting to all gift card module files
- Remove unused translations from CSV file
- Regenerate PHPStorm metadata for IDE support
- Ensure all files pass CI validation checks

Commit:
Fix remaining CI issues for gift card module

- Add missing 'Gift Card Code' translation
- Fix CSV file sorting and add newline at end
- Add #[\Override] attributes to all overridden methods
- Fix nullsafe operator usage
- Remove typed properties that cannot be typed in parent class
- Fix return types and method signatures for PHPStan compliance

Commit:
Fix duplicate Override attribute and include PHPStorm metadata

- Remove duplicate #[\Override] attribute in Payment/Info.php
- Include regenerated PHPStorm metadata files for IDE support

Commit:
Regenerate PHPStorm metadata without ProductFeeds module

- Remove ProductFeeds references from metadata
- Clean metadata now only includes core Maho modules and gift card module

mageaustralia committed:
Apply Rector code modernization fixes

- Remove redundant PHPDoc @param and @return tags (types already declared)
- Replace strpos() !== false with str_contains() for PHP 8.0+ compatibility
- Add missing #[\Override] attributes to overridden methods
- Optimize foreach loops using array_keys() where values not needed
- Modernize code to follow current PHP best practices

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
copyright update

fballiano committed:
Changed version number and consolidated installation script

fballiano committed:
Renamed entities from maho_giftcard to just giftcard

fballiano committed:
Slightly changing the product page layout

fballiano committed:
Improved backend experience

fballiano committed:
Improved product page layout

fballiano committed:
phpstorm + translations

fballiano committed:
Fixed PHPStan

trabulium commented:
> 😍😍😍😍😍 @trabulium can't wait to test it out!

@fballiano - I ran out of time last night trying to get it to pass all the tests, sorry about that. 

I really love the work you're doing and even just seeing your changes on this work now is why I wanted to see it open sourced. Having those extra eyes and extra input is important. I have more projects in the pipeline:

- graphQL API
- Point of Sales module
- The 404 / Redirect Manager, just needs porting across to core and fixing to pass tests etc
- A product feeds module
- Meilisearch

You did say you'd be happy the 404 / Redirect could become core. I think graphQL needs to become core also. I'm happy either way re: Point of Sale becoming core or not - As mentioned, the POS system uses Vue 3 with:

  - Pinia for state management
  - Vue Router for routing
  - Apollo Client for GraphQL API calls
  - Vite as the build tool
  
I'll see if there's a way I can make it more 'Maho' without the need for build tool and / or Vue etc but if I can't it may not be a good fit to become part of core. I'll update you as I go.

fballiano commented:
Thanks for the feedback @trabulium, it will take a bit more refinement (I'll keep working on it) because it's a complex and big module but it's such a nice addition!

About graphql, I was thinking we should explore API-platform framework to see if there's a way to implement it somehow in Maho. It would be even more amazing to be able to have all models automatically exposed via APIs, I don't know if it makes sense tho.

All the ideas you talk about are great, I'm just puzzled about the js stack for the point-of-sale, I've to say I would much rather prefer it to be HTML/CSS and nothing else, like another theme, if I'm not missing something important.

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano commented:
sorry, I wrote a bit quickly cause it's 11PM and I'm starting to see double 😅

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Removed Varien_

fballiano committed:
Fixed security, performance, and code quality issues in Gift Card module

Security fixes:
- Added CSRF protection via _setForcedFormKeyActions() in admin controller
- Fixed race condition in balance deduction with database transaction and row locking
- Removed inline onclick handlers in templates (CSP compliance)

Bug fixes:
- Fixed division by zero in Payment::refund() when $totalApplied is zero
- Fixed null pointer exception in _addHistory() when no admin session exists
- Added proper currency conversion in quote total calculation

Performance improvements:
- Added composite database indexes: (status, expires_at), (giftcard_id, created_at), (status, scheduled_at)

Code quality:
- Removed debug file_put_contents() logging from Total/Quote.php
- Refactored frontend JS to use module pattern (no global scope pollution)
- Used addEventListener instead of inline event handlers
- Wrapped hard-coded validation message in translation helper
- Used Mage_Core_Model_Locale::now() for timestamps

fballiano committed:
Moved Gift Cards menu from Customers to Sales section

fballiano committed:
Removed unused checkout/giftcards.phtml template

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Fixed translations: added missing, removed unused

fballiano committed:
work

fballiano committed:
Optimize gift card database structure: remove scheduled_email table

- Removed separate maho_giftcard_scheduled_email table
- Added email_scheduled_at and email_sent_at columns to maho_giftcard table
- Updated email sending to use core email queue (core_email_queue)
- Simplified cron job to query gift cards directly for pending emails
- Deleted Scheduled_Email model and resource classes
- Updated PHPStorm meta files

This reduces the database from 3 tables to 2 tables by merging the
1:1 scheduled email relationship into the main gift card table.
Gift card emails now flow through Maho's standard email infrastructure.

fballiano committed:
Simplify gift card configuration and make barcode package optional

- Move picqer/php-barcode-generator from require to suggest in composer.json
- Add frontend model to show warning when barcode package not installed
- Simplify system.xml: remove redundant comments, merge QR/barcode settings
- Remove unused send_on_creation config (auto-send when email provided)
- Auto-send gift card email only when recipient email is provided
- Clean up PDF template to use helper methods directly

fballiano committed:
lint

fballiano committed:
Add config-based defaults for gift card product attributes

- Add TextWithDefaultFromConfig form element for text fields that show
  system config default as placeholder when empty
- Update gift card attributes (lifetime, allow_message, is_redeemable)
  to use NULL = inherit from config pattern instead of hardcoded defaults
- Set attributes to visible=false so they render via custom tab only
- Add missing system config fields (allow_message, is_redeemable)
- Add helper methods to resolve product values vs config defaults
- Create custom product edit tab template with "Use Default" options
- Remove unused sales/order_address columns from install script
- Use frontend_type=boolean instead of select+yesno source model

fballiano committed:
Add missing translations for gift card module

fballiano committed:
lint

fballiano committed:
lint

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
work

fballiano committed:
we should have a "base_" currency system

fballiano committed:
we should have a "base_" currency system

fballiano committed:
work

fballiano committed:
work

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
rethought about base currency

fballiano committed:
work

fballiano committed:
work

fballiano committed:
work

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
work

fballiano committed:
work

fballiano committed:
Removed redeem, unused

fballiano committed:
check balance funcionality

fballiano committed:
check giftcard balance

fballiano committed:
work

fballiano committed:
Forbit buy a giftcard paying with a giftcard

fballiano committed:
Better admin fields management

fballiano committed:
Add giftcard to backend orders too

fballiano committed:
Add giftcard to backend orders too

fballiano committed:
lint

fballiano committed:
Fix gift card payment, refund, and credit memo handling

Payment fixes:
- Make Payment::capture() and Payment::refund() no-ops since gift card
  works as a discount, not a traditional payment method
- Balance deduction handled by Observer::deductGiftcardBalance()
- Balance restoration handled by Observer::refundGiftcardBalance()

Credit memo support:
- Enable credit memo for gift card orders by modifying core
  Order::canCreditmemo() to check for refundable gift card amounts
- For fully gift card paid orders, entire refund goes to gift card
- For partial gift card orders, proportional refund calculation
- Add setAllowZeroGrandTotal(true) to bypass validation when gift card
  covers entire refund
- Add sales_order_creditmemo_refund observer to restore gift card balance

Admin UI improvements:
- Fix gift card totals display in order/invoice/creditmemo views
- Use initTotals() pattern with proper block references
- Add "Update Totals" button to credit memo adjustments section
- Show positive amount with "Refund to Gift Card" label
- Add layout handles for updateqty AJAX responses

Total collectors:
- Fix ordering with after="subtotal,shipping,discount,tax,grand_total"
- Include tax in proportional calculations for partial orders

Other fixes:
- Add missing admin payment info/form templates
- Fix quote-to-order conversion for backend orders
- Remove separate transaction from deductGiftcardBalance() for atomicity

fballiano committed:
Fix PHPStan errors in giftcard blocks

Add PHPDoc type hints for parent blocks in totals classes and fix
return value in Payment/Info block to match parent class behavior.

fballiano committed:
lint

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Add gift card cart item renderer with localized date formatting

- Create custom cart item renderer for gift card products
- Format ISO dates (YYYY-MM-DD) to localized long format at view time
- Centralize display options in helper with DISPLAY_OPTIONS constant
- Use buildAdditionalOptions() helper method in product type and observer

fballiano committed:
Updating a card already in the cart

fballiano committed:
lint

fballiano committed:
Added tests

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
lint

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
new phpstangs

fballiano committed:
work

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
copy

fballiano committed:
tests

fballiano committed:
Reordering

fballiano committed:
Reviewed price templates

fballiano committed:
Reviewed price templates

fballiano committed:
work

fballiano committed:
work

fballiano committed:
work

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
work

fballiano committed:
work

fballiano committed:
work

fballiano committed:
work

fballiano committed:
work

fballiano committed:
work

fballiano committed:
work

fballiano committed:
Composer update

fballiano committed:
Merge main branch

fballiano committed:
Restructured claude.md

fballiano committed:
Merge branch 'main' into feature/gift-card-module

fballiano committed:
Improve gift card checkout UX and remove unused payment method

Checkout improvements:
- Hide non-zero-total payment methods (checkmo, etc.) when gift cards fully cover order
- Remove redundant "Gift Card" payment method section
- Use standard "free" payment method for zero-total orders

Admin form improvements:
- Simplify new gift card creation: single "Amount" field instead of redundant Balance/Initial Balance
- Edit mode shows "Initial Balance" (read-only) and "Current Balance" (editable) with clear helper text
- Automatically set initial_balance = balance on creation

Gift card status:
- Fix status update to "used" when balance reaches $0.00

Code cleanup:
- Remove unused gift card payment method (giftcard was never available in checkout)
- Remove obsolete payment form/info blocks and templates
- Remove payment helper override and checkout methods block override
- Remove auto-select payment method observer
- Remove payment method configuration from config.xml
- Update PHPStorm metadata

Gift cards now work purely as a discount mechanism that reduces order total.
When fully covered, the standard "free" payment method is used.

fballiano committed:
work

fballiano committed:
work

fballiano committed:
work

fballiano committed:
Fix gift card checkout payment method handling and remove obsolete tests

This commit fixes the checkout UX when gift cards fully cover orders and
removes the obsolete payment method test that was testing non-existent functionality.

Payment Method Handling:
- Fixed Observer to properly enable (not disable) the "free" payment method
  when gift cards fully cover the order (grand total = $0)
- Added server-side payment methods HTML generation in CartController
- Modified AJAX responses to include fresh payment methods HTML when in
  checkout context (checkout=1 parameter)
- Updated frontend JavaScript to receive and render server-provided payment
  methods HTML, ensuring the payment section stays in sync with the quote state

Test Cleanup:
- Removed tests/Backend/Integration/Giftcard/Model/PaymentTest.php
- This test was for a dedicated "giftcard" payment method that was removed
- The current implementation uses the standard "free" payment method when
  gift cards fully cover orders, not a separate payment method
- All 1120 tests now pass (89,547 assertions)

The changes ensure that when customers apply/remove gift cards during checkout,
the payment methods section properly updates to show only the "free" method
when the order is fully covered, and restores other payment methods when needed.

fballiano commented:
To me this is finished, i've remove the giftcard payment method, it now uses the started "no payment" if the giftcards cover the whole order. I think it's better and simplifies the code of this PR.

I tested everything I could multiple times and to me this is good to merge

fballiano committed:
lint

fballiano committed:
refund for canceled orders

fballiano committed:
Fix unique constraint violation in order cancellation tests

Use uniqid() for order increment IDs instead of hardcoded values
to prevent UNIQUE constraint violations when tests run in parallel
or after other test suites.

trabulium commented:
Hi @fballiano  - Amazing to see this merged! Sorry, I didn't see your earlier message 3 days ago. I've just pulled down latest main and am testing it today. I feel a little guilty you had to put in so much work to bring it up to the Maho standard. I spent more time on the Feed Extension to ensure better code quality and hopefully a lot less work for you. I've imported all our GC's now and pagination between them is incredibly fast in Maho @ ~85-100ms. I've found one small bug I'll push a PR for later - I'll just test a bit more and see if I can find any more.
